File: packages/shared-backend/payments/razorpay.py
"""
Razorpay client and webhook utilities
"""

import logging
import razorpay
from typing import Optional
from ..config.base import get_settings

log = logging.getLogger(__name__)

# Global Razorpay client (lazy initialization)
_razorpay_client: Optional[razorpay.Client] = None


def get_razorpay_client() -> Optional[razorpay.Client]:
    """
    Get or create Razorpay client instance.
    
    Returns:
        Razorpay client or None if not configured
    """
    global _razorpay_client
    
    if _razorpay_client is not None:
        return _razorpay_client
    
    try:
        settings = get_settings()
        key_id = settings.razorpay_key_id if hasattr(settings, 'razorpay_key_id') else None
        key_secret = settings.razorpay_key_secret if hasattr(settings, 'razorpay_key_secret') else None
    except Exception:
        # Fallback to environment variables
        # Note: This shared package uses unprefixed variables for cross-app compatibility
        # Apps should set prefixed variables (ASK_RAZORPAY_KEY_ID, etc.) in their .env files
        import os
        key_id = os.getenv("RAZORPAY_KEY_ID") or os.getenv("LIVE_KEY_ID")
        key_secret = os.getenv("RAZORPAY_KEY_SECRET") or os.getenv("LIVE_KEY_SECRET")
    
    if not key_id or not key_secret:
        log.warning("Razorpay credentials not configured")
        return None
    
    try:
        _razorpay_client = razorpay.Client(auth=(key_id, key_secret))
        return _razorpay_client
    except Exception as e:
        log.error(f"Failed to initialize Razorpay client: {e}")
        return None


def verify_webhook_signature(payload: str, signature: str, webhook_secret: Optional[str] = None) -> bool:
    """
    Verify Razorpay webhook signature.
    
    Args:
        payload: Webhook payload string
        signature: Webhook signature from header
        webhook_secret: Optional webhook secret (uses settings if not provided)
        
    Returns:
        True if signature is valid, False otherwise
    """
    client = get_razorpay_client()
    if not client:
        return False
    
    try:
        settings = get_settings()
        secret = webhook_secret or (getattr(settings, 'RAZORPAY_WEBHOOK_SECRET', None) if hasattr(settings, 'RAZORPAY_WEBHOOK_SECRET') else None)
    except Exception:
        # Fallback to environment variable
        # Note: This shared package uses unprefixed variables for cross-app compatibility
        import os
        secret = webhook_secret or os.getenv("RAZORPAY_WEBHOOK_SECRET")
    
    if not secret:
        log.warning("Razorpay webhook secret not configured")
        return False
    
    try:
        client.utility.verify_webhook_signature(
            payload,
            signature,
            secret
        )
        return True
    except Exception as e:
        log.error(f"Webhook signature verification failed: {e}")
        return False

File: packages/shared-frontend/src/payments/razorpay.ts
/**
 * Razorpay client for payment and subscription management
 * Shared across KVSHVL platform applications
 * 
 * Note: This shared package uses unprefixed environment variables (RAZORPAY_KEY_ID, etc.)
 * for cross-app compatibility. Individual apps should set prefixed variables
 * (SKETCH2BIM_RAZORPAY_KEY_ID, etc.) in their .env files,
 * but the code checks unprefixed variables first for shared functionality.
 */

import Razorpay from "razorpay";
// Note: Using console methods in shared package since logger is not available

/**
 * Razorpay client instance (lazy initialization)
 */
let razorpayInstance: Razorpay | null = null;

/**
 * Get the Razorpay client instance (lazy initialization)
 * Only creates the client when first called, not during module import
 */
export function getRazorpayClient(): Razorpay {
  if (!razorpayInstance) {
    // Check unprefixed variables first (shared across all apps)
    // Apps can set prefixed variables (SKETCH2BIM_RAZORPAY_KEY_ID, etc.) but
    // this shared package uses unprefixed for compatibility
    const keyId = process.env.RAZORPAY_KEY_ID;
    const keySecret = process.env.RAZORPAY_KEY_SECRET;

    if (!keyId || !keySecret) {
      throw new Error("Razorpay credentials not configured. Please set RAZORPAY_KEY_ID and RAZORPAY_KEY_SECRET environment variables.");
    }

    razorpayInstance = new Razorpay({
      key_id: keyId,
      key_secret: keySecret,
    });
  }
  return razorpayInstance;
}

/**
 * Verify Razorpay webhook signature
 * @param payload - Raw webhook payload string
 * @param signature - Webhook signature from x-razorpay-signature header
 * @returns true if signature is valid
 */
export function verifyWebhookSignature(payload: string, signature: string): boolean {
  const webhookSecret = process.env.RAZORPAY_WEBHOOK_SECRET;
  
  if (!webhookSecret) {
    console.error("RAZORPAY_WEBHOOK_SECRET not configured");
    return false;
  }

  try {
    const crypto = require("crypto");
    const expectedSignature = crypto
      .createHmac("sha256", webhookSecret)
      .update(payload)
      .digest("hex");

    return expectedSignature === signature;
  } catch (error) {
    console.error("Webhook signature verification failed:", error);
    return false;
  }
}

